In [1]:
import sys
import os
sys.path.append(os.path.abspath('..'))

In [2]:
import pandas as pd
from src.cleaning import extract_category, concat_dataset, clean_sold_number, normalize_material_name, normalize_brand, flatten_dataset, check_null_rate, parse_warranty, apply_warranty, parse_size, parse_thickness, normalize_firmness, check_empty_variations, load_and_clean_dataset, standardize_origin, clean_text
from src.data_loader import load_data_raw   

In [3]:
vuanem_dataset = load_data_raw("vuanem.json")
vuanem_dataset["category"] = vuanem_dataset["material_type"].apply(extract_category)
vuanem_dataset["product_sold_number"] = vuanem_dataset["product_sold_number"].apply(clean_sold_number)
display(vuanem_dataset.head())

,product_name,image_url,link,brand,product_sold_number,rating,reviews,description,specifications,variations,material_type,category
0,Nệm cao su thiên nhiên Gummi Classic thế hệ mớ...,"https://cdn-cf-ops.vuanem.com/186x186,webp,q80...",https://vuanem.com/nem-cao-su-gummi-classic.ht...,Gummi,44410,4.9,359,"Thương hiệu Gummi - Tinh khiết từ cao su, đúng...","{'origin': 'Việt Nam', 'warranty': 180, 'layer...","[{'size': '120x200', 'thickness': '5', 'price'...",Cao su thiên nhiên,Cao su
1,Nệm lò xo Amando Elite Original túi độc lập ti...,"https://cdn-cf-ops.vuanem.com/186x186,webp,q80...",https://vuanem.com/nem-lo-xo-amando-elite-orig...,Amando,26658,5.0,273,Thương hiệu Amando – Nghệ thuật giấc ngủ châu ...,"{'origin': 'Việt Nam', 'warranty': 120, 'layer...","[{'size': '120x200', 'thickness': '23', 'price...",Nệm Hybrid (Lò xo túi độc lập + Foam),Hybrid
2,Nệm foam Comfy Cloud 2.0 siêu đàn hồi dày 15cm,"https://cdn-cf-ops.vuanem.com/186x186,webp,q80...",https://vuanem.com/nem-foam-comfy-cloud-20.htm...,Comfy,25068,4.9,278,Comfy – Khoa học giấc ngủ cho cuộc sống cân bằ...,"{'origin': 'Việt Nam', 'warranty': 120, 'layer...","[{'size': '120x200', 'thickness': '15', 'price...",Nệm Hybrid (Cao su thiên nhiên + Foam),Hybrid
3,Nệm foam công nghệ Đức Goodnight Active Hybrid...,"https://cdn-cf-ops.vuanem.com/186x186,webp,q80...",https://vuanem.com/nem-foam-goodnight-active-h...,Goodnight,23094,5.0,139,Goodnight Active Hybrid - Chiếc nệm ĐẦU TIÊN ở...,"{'origin': 'Việt Nam', 'warranty': 120, 'layer...","[{'size': '120x200', 'thickness': '20', 'price...",Nệm Hybrid (Cao su thiên nhiên + Foam),Hybrid
4,Nệm lò xo Goodnight SleepWave Hybrid (Sleep Wa...,"https://cdn-cf-ops.vuanem.com/186x186,webp,q80...",https://vuanem.com/nem-lo-xo-goodnight-sleepwa...,Goodnight,22162,4.7,176,Goodnight SleepWave Hybrid – Công nghệ Hybrid ...,"{'origin': 'Việt Nam', 'warranty': 120, 'layer...","[{'size': '120x200', 'thickness': '20', 'price...",Nệm Hybrid (Lò xo túi độc lập + Foam),Hybrid


In [4]:
df_empty = check_empty_variations(
    vuanem_dataset, 
    "vuanem"
)

=== vuanem ===
Tổng sản phẩm:              57
Sản phẩm null/rỗng variations: 0
Tỉ lệ:                      0.0%


In [5]:
khonemtonghop_dataset = load_data_raw("khonemtonghop.json")
display(khonemtonghop_dataset.head())

,product_name,image_url,link,product_sold_number,rating,reviews,category,description,variations,brand,specifications,material_type
0,Nệm cao su thiên nhiên Beetex Cool Pure,https://khonemtonghop.com/wp-content/uploads/2...,https://khonemtonghop.com/nem-cao-su-thien-nhi...,32,5.0,3,Cao su thiên nhiên,Nội dung:\nThông số chi tiết của nệm cao su th...,"[{'size': '100x200', 'thickness': '10cm', 'pri...",Beetex,"{'origin': 'Việt Nam', 'warranty': 144, 'layer...",Cao su thiên nhiên
1,Nệm Cao Su Thiên Nhiên Beetex Pure,https://khonemtonghop.com/wp-content/uploads/2...,https://khonemtonghop.com/nem-cao-su-thien-nhi...,23,5.0,14,Cao su thiên nhiên,Nội dung:\nThông số chi tiết của nệm cao su th...,"[{'size': '100x200', 'thickness': '10cm', 'pri...",Beetex,"{'origin': 'Việt Nam', 'warranty': 144, 'layer...",Cao su thiên nhiên
2,Nệm cao su thiên nhiên Beetex Cool Massage,https://khonemtonghop.com/wp-content/uploads/2...,https://khonemtonghop.com/nem-cao-su-thien-nhi...,49,5.0,29,Cao su thiên nhiên,Nội dung:\nNệm cao su thiên nhiên Beetex Cool ...,"[{'size': '120x200', 'thickness': '10cm', 'pri...",Beetex,"{'origin': 'Việt Nam', 'warranty': 180, 'layer...",Cao su thiên nhiên
3,Nệm cao su thiên nhiên Beetex Latex,https://khonemtonghop.com/wp-content/uploads/2...,https://khonemtonghop.com/nem-cao-su-thien-nhi...,92,5.0,30,Cao su thiên nhiên,Nội dung:\nGiới thiệu về nệm cao su thiên nhiê...,"[{'size': '100x200', 'thickness': '10cm', 'pri...",Beetex,"{'origin': 'Việt Nam', 'warranty': 144, 'layer...",Cao su thiên nhiên
4,Nệm Cao Su Nhân Tạo Beetex Foam,https://khonemtonghop.com/wp-content/uploads/2...,https://khonemtonghop.com/nem-cao-su-nhan-tao-...,23,4.9,7,Cao su thiên nhiên,Nội dung:\nGiới thiệu nệm cao su nhân tạo Beet...,"[{'size': '100x200', 'thickness': '10cm', 'pri...",Beetex,"{'origin': 'Việt Nam', 'warranty': 96, 'layer_...",Cao su nhân tạo


In [6]:
df_empty = check_empty_variations(
    khonemtonghop_dataset, 
    "khonemtonghop"
)

=== khonemtonghop ===
Tổng sản phẩm:              74
Sản phẩm null/rỗng variations: 0
Tỉ lệ:                      0.0%


In [7]:
thegioinem_dataset = load_data_raw("deals-thegioinem.json")
thegioinem_dataset['category'] = thegioinem_dataset["material_type"].apply(extract_category)
thegioinem_dataset.rename(columns={'review': 'reviews'}, inplace=True)
display(thegioinem_dataset.head())

,product_name,image_url,link,product_sold_number,rating,reviews,description,brand,specifications,variations,material_type,category
0,Nệm cao su thiên nhiên than hoạt tính TATANA C...,https://thegioinem.com/upload/thumb/202510/09/...,https://thegioinem.com/nem-cao-su-thien-nhien-...,335,5.0,275,FEEDBACK CỦA KHÁCH HÀNG\nKHI TRẢI NGHIỆM 102 N...,TATANA,"{'origin': 'Việt Nam', 'warranty': '10 năm (bở...","[{'size': '100x200', 'thickness': '10', 'price...",Cao su thiên nhiên,Cao su
1,Nệm cao su thiên nhiên TATANA Cool Fresh,https://thegioinem.com/upload/thumb/202411/07/...,https://thegioinem.com/nem-cao-su-thien-nhien-...,553,5.0,334,FEEDBACK CỦA KHÁCH HÀNG\nKHI TRẢI NGHIỆM 102 N...,TATANA,"{'origin': 'Việt Nam', 'warranty': '10 năm (bở...","[{'size': '100x200', 'thickness': '10', 'price...",Cao su thiên nhiên,Cao su
2,Nệm cao su thiên nhiên TATANA Standard,https://thegioinem.com/upload/thumb/202411/07/...,https://thegioinem.com/nem-cao-su-thien-nhien-...,675,5.0,197,FEEDBACK CỦA KHÁCH HÀNG\nKHI TRẢI NGHIỆM 102 N...,TATANA,"{'origin': 'Việt Nam', 'warranty': '10 năm (bở...","[{'size': '100x200', 'thickness': '5', 'price'...",Cao su thiên nhiên,Cao su
3,Nệm cao su thiên nhiên Vạn Thành Standard,https://thegioinem.com/upload/thumb/202506/19/...,https://thegioinem.com/nem-cao-su-van-thanh-st...,4500,5.0,88,Nệm cao su Vạn Thành Standard là dòng sản phẩm...,Vạn Thành,"{'origin': 'Việt Nam', 'warranty': '12 năm', '...","[{'size': '100x200', 'thickness': '5', 'price'...",Cao su thiên nhiên,Cao su
4,Nệm cao su thiên nhiên Vạn Thành Unique,https://thegioinem.com/upload/thumb/202506/19/...,https://thegioinem.com/nem-cao-su-van-thanh-un...,3300,5.0,99,Nệm cao su Vạn Thành Unique được sản xuất từ c...,Vạn Thành,"{'origin': 'Việt Nam', 'warranty': '15 năm', '...","[{'size': '140x200', 'thickness': '10', 'price...",Cao su tổng hợp,Foam


In [8]:
df_empty = check_empty_variations(
    thegioinem_dataset, 
    "thegioinem"
)

=== thegioinem ===
Tổng sản phẩm:              159
Sản phẩm null/rỗng variations: 0
Tỉ lệ:                      0.0%


In [9]:
thegioigiuongnem_dataset = load_data_raw("deals-thegioigiuongnem.json")
thegioigiuongnem_dataset.rename(columns={'review': 'reviews'}, inplace=True)
display(thegioigiuongnem_dataset.head())

,product_name,image_url,link,product_sold_number,rating,reviews,description,brand,specifications,variations,material_type,category
0,Nệm Cao Su Thiên Nhiên Vạn Thành Standard,https://bizweb.dktcdn.net/thumb/medium/100/405...,https://thegioigiuongnem.com/nem-cao-su-thien-...,NaN,NaN,NaN,Bảng Giá Nệm Cao Su Thiên Nhiên Vạn Thành Stan...,VẠN THÀNH,"{'origin': 'Việt Nam', 'warranty': 'Bảo Hành: ...","[{'size': '100x200', 'thickness': '05', 'price...",Cao su tổng hợp,Foam
1,Nệm Cao Su Thiên Nhiên Kim Cương Happy Gold,https://bizweb.dktcdn.net/thumb/medium/100/405...,https://thegioigiuongnem.com/nem-cao-su-thien-...,NaN,NaN,NaN,Bảng Giá Nệm Cao Su Thiên Nhiên Kim Cương Happ...,KIM CƯƠNG,"{'origin': 'Việt Nam', 'warranty': 'BẢO HÀNH: ...","[{'size': '100x200', 'thickness': '05', 'price...",Cao su thiên nhiên,Cao su
2,NỆM CAO SU LIÊN Á - HYBRID IKI EZ,https://bizweb.dktcdn.net/thumb/medium/100/405...,https://thegioigiuongnem.com/nem-cao-su-lien-a...,NaN,NaN,NaN,NỆM CAO SU LIÊN Á - HYBRID IKI EZ\nVới sự kết ...,LIÊN Á,"{'origin': 'Việt Nam', 'warranty': 'Bảo Hành: ...","[{'size': '100x200', 'thickness': '10', 'price...",Nệm Hybrid (Foam + Cao su thiên nhiên),Hybrid
3,Đệm Cao Su Everon - Natural Charcoal Latex,https://bizweb.dktcdn.net/thumb/medium/100/405...,https://thegioigiuongnem.com/dem-cao-su-everon...,NaN,NaN,NaN,Thông tin đệm cao su Everon - Natural Charcoal...,EVERON,"{'origin': 'Việt Nam', 'warranty': 'Bảo Hành: ...","[{'size': '120x200', 'thickness': '10', 'price...",Cao su thiên nhiên,Cao su
4,Nệm Cao Su Liên Á - 5Zone Premium | Nâng Đỡ Tố...,https://bizweb.dktcdn.net/thumb/medium/100/405...,https://thegioigiuongnem.com/nem-cao-su-lien-a...,NaN,NaN,NaN,Nệm cao su Liên Á 5Zone Premium - Nâng đỡ tối ...,NỆM LIÊN Á,"{'origin': 'Việt Nam', 'warranty': 'bảo hành 1...","[{'size': '160x200', 'thickness': '15', 'price...",Cao su,Cao su


In [10]:
df_empty = check_empty_variations(
    thegioigiuongnem_dataset, 
    "thegioigiuongnem"
)

=== thegioigiuongnem ===
Tổng sản phẩm:              163
Sản phẩm null/rỗng variations: 11
Tỉ lệ:                      6.7%

Danh sách sản phẩm bị lỗi:
                                         product_name         brand
0   Nệm Cao Su Dunlopillo - Latex World Neo 160x20...    DUNLOPILLO
1                Gối Ôm Bông Gòn Cao Cấp Hàn Việt Hải  HÀN VIỆT HẢI
2                      Nệm Bông Ép Gấp 03 - Vạn Thành     VẠN THÀNH
3             Nệm Kim Cương Bông Ép - Eufiber Classic     KIM CƯƠNG
4      Nệm Bông Ép Cao Cấp  Emerald - Bảo Hành 15 Năm         KOREA
5                  Nệm Bông Ép Liên Á - Mliving Fiber    NỆM LIÊN Á
6                       Nệm Lò Xo Túi Edena - Premium         EDENA
7                           Nệm Lò Xo Diamond - Edena         EDENA
8                     Nệm Lò Xo Vạn Thành - Excellent     VẠN THÀNH
9   Diamond - Nệm Lò Xo Túi Độc Lập Lót Cao Su Thi...     VẠN THÀNH
10              Nệm Lò Xo Cao Cấp Vạn Thành - Perfect     VẠN THÀNH


In [11]:
thegioigiuongnem_dataset = load_and_clean_dataset(
    "deals-thegioigiuongnem.json",
    thegioigiuongnem_dataset
)

Đã xóa 11 sản phẩm không có variations
Còn lại: 152 sản phẩm


In [12]:
df_empty = check_empty_variations(
    thegioigiuongnem_dataset, 
    "thegioigiuongnem"
)

=== thegioigiuongnem ===
Tổng sản phẩm:              152
Sản phẩm null/rỗng variations: 0
Tỉ lệ:                      0.0%


In [13]:
mattress_dataset = concat_dataset(
    vuanem_dataset, 
    thegioinem_dataset, 
    khonemtonghop_dataset, 
    thegioigiuongnem_dataset
)

In [14]:
display(mattress_dataset.head())

,product_name,image_url,link,brand,product_sold_number,rating,reviews,description,specifications,variations,material_type,category
0,Nệm cao su thiên nhiên Gummi Classic thế hệ mớ...,"https://cdn-cf-ops.vuanem.com/186x186,webp,q80...",https://vuanem.com/nem-cao-su-gummi-classic.ht...,Gummi,44410.0,4.9,359.0,"Thương hiệu Gummi - Tinh khiết từ cao su, đúng...","{'origin': 'Việt Nam', 'warranty': 180, 'layer...","[{'size': '120x200', 'thickness': '5', 'price'...",Cao su thiên nhiên,Cao su
1,Nệm lò xo Amando Elite Original túi độc lập ti...,"https://cdn-cf-ops.vuanem.com/186x186,webp,q80...",https://vuanem.com/nem-lo-xo-amando-elite-orig...,Amando,26658.0,5.0,273.0,Thương hiệu Amando – Nghệ thuật giấc ngủ châu ...,"{'origin': 'Việt Nam', 'warranty': 120, 'layer...","[{'size': '120x200', 'thickness': '23', 'price...",Nệm Hybrid (Lò xo túi độc lập + Foam),Hybrid
2,Nệm foam Comfy Cloud 2.0 siêu đàn hồi dày 15cm,"https://cdn-cf-ops.vuanem.com/186x186,webp,q80...",https://vuanem.com/nem-foam-comfy-cloud-20.htm...,Comfy,25068.0,4.9,278.0,Comfy – Khoa học giấc ngủ cho cuộc sống cân bằ...,"{'origin': 'Việt Nam', 'warranty': 120, 'layer...","[{'size': '120x200', 'thickness': '15', 'price...",Nệm Hybrid (Cao su thiên nhiên + Foam),Hybrid
3,Nệm foam công nghệ Đức Goodnight Active Hybrid...,"https://cdn-cf-ops.vuanem.com/186x186,webp,q80...",https://vuanem.com/nem-foam-goodnight-active-h...,Goodnight,23094.0,5.0,139.0,Goodnight Active Hybrid - Chiếc nệm ĐẦU TIÊN ở...,"{'origin': 'Việt Nam', 'warranty': 120, 'layer...","[{'size': '120x200', 'thickness': '20', 'price...",Nệm Hybrid (Cao su thiên nhiên + Foam),Hybrid
4,Nệm lò xo Goodnight SleepWave Hybrid (Sleep Wa...,"https://cdn-cf-ops.vuanem.com/186x186,webp,q80...",https://vuanem.com/nem-lo-xo-goodnight-sleepwa...,Goodnight,22162.0,4.7,176.0,Goodnight SleepWave Hybrid – Công nghệ Hybrid ...,"{'origin': 'Việt Nam', 'warranty': 120, 'layer...","[{'size': '120x200', 'thickness': '20', 'price...",Nệm Hybrid (Lò xo túi độc lập + Foam),Hybrid


In [15]:
mattress_dataset['material_type'] = mattress_dataset['material_type'].apply(normalize_material_name)

In [16]:
print(mattress_dataset['brand'].nunique())         
print(mattress_dataset['material_type'].nunique()) 
print(mattress_dataset['product_name'].nunique()) 

43
37
430


In [17]:
# Xem các hãng nệm nào đang chiếm đa số
print(mattress_dataset['brand'].value_counts())

brand
Dunlopillo            40
TATANA                39
Kim Cương             37
Vạn Thành             33
KIM CƯƠNG             29
Liên Á                23
DUNLOPILLO            19
HÀN VIỆT HẢI          15
LIÊN Á                14
Edena                 12
KOREA                 12
Daafar                12
Comfy                 11
EDENA                 11
ROMANTIC              10
VẠN THÀNH             10
King Koil             10
Beetex                10
Everon                 9
Gummi                  7
Goodnight              7
EVERON                 6
Amando                 6
USA GOLDEN BEDDING     5
Ưu Việt                5
NỆM LIÊN Á             5
Aeroflow               4
khác                   4
ĐỒNG PHÚ               4
ACB Pro                4
Wonjun                 4
THẮNG LỢI              3
Tempur                 3
Spring Air             3
Đồng Phú               3
VINAMATTRESS           3
Bedgear                2
SPRING AIR             2
Oyasumi                2
Hàn Việt Hải       

- Có những sample cùng 1 brand nhưng tên của các brand khác nhau về mặt kí tự (như chữ viết hoa, viết sai chính tả) -> Cần chuẩn hóa để chúng cùng 1 định dạng chung và về cùng 1 brand duy nhất.

In [18]:
mattress_dataset['brand'] = mattress_dataset['brand'].apply(normalize_brand)
print(f"Số brand sau chuẩn hoá: {mattress_dataset['brand'].nunique()}")
print(mattress_dataset['brand'].value_counts())

Số brand sau chuẩn hoá: 31
brand
Kim Cương             66
Dunlopillo            59
Vạn Thành             43
Liên Á                42
Tatana                39
Edena                 23
Hàn Việt Hải          16
Everon                15
Daafar                12
Korea                 12
Romantic              11
Comfy                 11
Beetex                10
King Koil             10
Gummi                  7
Goodnight              7
Đồng Phú               7
Amando                 6
Ưu Việt                5
Spring Air             5
Usa Golden Bedding     5
Aeroflow               4
Wonjun                 4
Acb Pro                4
Tempur                 3
Thắng Lợi              3
Vinamattress           3
Bedgear                2
Oyasumi                2
Vivian                 1
Double Win             1
Name: count, dtype: int64


In [19]:
print(mattress_dataset['material_type'].value_counts())

material_type
Cao su thiên nhiên                                                      82
Bông ép                                                                 60
Foam                                                                    40
Nệm Hybrid (Foam + Lò xo túi độc lập)                                   33
Lò xo túi độc lập                                                       24
Cao su tổng hợp                                                         21
Nệm Hybrid (Cao su thiên nhiên + Foam)                                  17
Lò xo liên kết                                                          17
Nệm Hybrid (Cao su tổng hợp + Foam)                                     17
Nệm Hybrid (Cao su thiên nhiên + Lò xo túi độc lập)                     14
Nệm Hybrid (Bông ép + Foam)                                             13
Lò xo                                                                   12
Nệm Hybrid (Cao su thiên nhiên + Foam + Lò xo túi độc lập)              11
Cao su     

*Các trường hợp cần được xử lý* 
- Lò xo liên kết, Lò xo normablock, Lò lo liên kết -> Lò xo liên kết
- Lò xo túi độc lập, Lò lo túi độc lập, Lò lo túi, Lò xo độc lập, Lò lo túi liên kết, Lò lo túi độc lập Norm Acitve -> Lò xo túi độc lập
- Bông ép, Bông nhân tạo -> Bông ép
- Nệm Hybrid (Foam) -> Foam
- Nệm Hybrid (Cao su thiên nhiên + Lò xo túi độc lập + Foam),  Nệm Hybrid (Foam + Lò xo túi độc lập + Cao su thiên nhiên)
-  Nệm Hybrid (Cao su + Lò xo túi độc lập + Foam), Nệm Hybrid (Lò xo túi độc lập + cao su + Foam)
- Nệm Hybrid (Foam + Cao su thiên nhiên), Nệm Hybrid (Cao su thiên nhiên + Foam)

In [20]:
mattress_dataset['category'] = mattress_dataset["material_type"].apply(extract_category)

In [21]:
rows = []
for _, product in mattress_dataset.iterrows():
    base = {
        "product_name":  product.get("product_name"),
        "brand":         product.get("brand"),
        "material_type": product.get("material_type"),
        "category":      product.get("category"),
        "description":   product.get("description"),
        "rating":        product.get("rating"),
        "firmness":      product.get("specifications", {}).get("firmness") 
                         if isinstance(product.get("specifications"), dict) else None,
    }
    for var in product.get("variations", []):
        row = base.copy()
        row["size"]      = var.get("size")
        row["thickness"] = var.get("thickness")
        row["price"]     = var.get("price")
        rows.append(row)

df_flat = pd.DataFrame(rows)
print(df_flat.shape)
print(df_flat['price'].describe())

(4532, 10)
count    4.532000e+03
mean     9.669327e+06
std      1.179285e+07
min      7.200000e+04
25%      3.192000e+06
50%      5.887000e+06
75%      1.146700e+07
max      1.576800e+08
Name: price, dtype: float64


In [22]:
print(mattress_dataset.groupby('category')['product_name'].count())
print(mattress_dataset.groupby('brand')['product_name'].count())

category
Bông ép     60
Cao su      91
Foam        61
Hybrid     175
Khác         2
Lò xo       53
Name: product_name, dtype: int64
brand
Acb Pro                4
Aeroflow               4
Amando                 6
Bedgear                2
Beetex                10
Comfy                 11
Daafar                12
Double Win             1
Dunlopillo            59
Edena                 23
Everon                15
Goodnight              7
Gummi                  7
Hàn Việt Hải          16
Kim Cương             66
King Koil             10
Korea                 12
Liên Á                42
Oyasumi                2
Romantic              11
Spring Air             5
Tatana                39
Tempur                 3
Thắng Lợi              3
Usa Golden Bedding     5
Vinamattress           3
Vivian                 1
Vạn Thành             43
Wonjun                 4
Đồng Phú               7
Ưu Việt                5
Name: product_name, dtype: int64


In [23]:
# Kiểm tra outlier giá
print(df_flat[df_flat['price'] > 50000000][['product_name', 'brand', 'price']])

                                           product_name       brand     price
132   Nệm lò xo Dunlopillo de.Stress Release túi độc...  Dunlopillo  52464000
206   Nệm foam Nhật Bản Aeroflow Pride bảo vệ cột số...    Aeroflow  51799000
207   Nệm foam Nhật Bản Aeroflow Pride bảo vệ cột số...    Aeroflow  51799000
211   Nệm cao su cao cấp Gummi Legend nhập khẩu dày ...       Gummi  53699000
232   Nệm lò xo Bedgear H2 Hybrid thoáng khí kép dày...     Bedgear  51350000
...                                                 ...         ...       ...
4210  Royal Kensington - Nệm Lò Xo Túi Cao Cấp Lót C...  Dunlopillo  69444000
4211  Royal Kensington - Nệm Lò Xo Túi Cao Cấp Lót C...  Dunlopillo  74801000
4218  Elizabeth - Nệm Lò Xo Túi Cao Cấp Lót Cao Su D...  Dunlopillo  55655000
4219  Elizabeth - Nệm Lò Xo Túi Cao Cấp Lót Cao Su D...  Dunlopillo  58379000
4220  Elizabeth - Nệm Lò Xo Túi Cao Cấp Lót Cao Su D...  Dunlopillo  59278000

[76 rows x 3 columns]


In [24]:
print(mattress_dataset['category'].value_counts())

category
Hybrid     175
Cao su      91
Foam        61
Bông ép     60
Lò xo       53
Khác         2
Name: count, dtype: int64


In [25]:
# Xem từ mattress_dataset gốc
mattress_dataset['specifications'].apply(
    lambda x: x.get('warranty') if isinstance(x, dict) else None
).value_counts()

specifications
120                            72
10 năm                         57
Bảo Hành: 10 Năm               41
10 năm (bởi công ty TATANA)    31
5 năm                          16
                               ..
Bảo hành 08 năm                 1
BẢO HÀNH: 18 NĂM                1
BẢO HÀNH : 05 NĂM               1
BẢO HÀNH: 06 NĂM                1
Bảo Hành: 7 Năm                 1
Name: count, Length: 67, dtype: int64

## Chuẩn hoá cột Warranty

### Vấn đề
Cột `warranty` trong raw data có nhiều định dạng khác nhau:
- Số thuần: `120` (đã chuẩn hoá khi crawl từ vuanem, khonemtonghop)
- Chuỗi có đơn vị năm: `"10 năm"`, `"Bảo Hành: 10 Năm"`, `"BẢO HÀNH 10 NĂM"`
- Chuỗi phức tạp: `"10 năm (bởi công ty TATANA)"`, `"6-15 năm"`

### Quyết định xử lý
- Quy đổi tất cả về **số tháng** (float)
- Trường hợp khoảng "6-15 năm" → lấy số đầu tiên (bảo hành tối thiểu)
- Không thể parse → `None`

In [26]:
mattress_dataset['warranty'] = mattress_dataset[
    'specifications'
].apply(apply_warranty)

# Kiểm tra kết quả
print(mattress_dataset['warranty'].value_counts().sort_index())
print(f"\nNull rate: {mattress_dataset['warranty'].isna().sum() / len(mattress_dataset) * 100:.1f}%")

warranty
12.0       1
36.0       3
60.0      61
72.0       5
84.0      15
96.0      22
108.0      1
120.0    233
144.0     60
180.0     30
216.0      2
240.0      1
Name: count, dtype: int64

Null rate: 1.8%


In [27]:
display(mattress_dataset.head())

,product_name,image_url,link,brand,product_sold_number,rating,reviews,description,specifications,variations,material_type,category,warranty
0,Nệm cao su thiên nhiên Gummi Classic thế hệ mớ...,"https://cdn-cf-ops.vuanem.com/186x186,webp,q80...",https://vuanem.com/nem-cao-su-gummi-classic.ht...,Gummi,44410.0,4.9,359.0,"Thương hiệu Gummi - Tinh khiết từ cao su, đúng...","{'origin': 'Việt Nam', 'warranty': 180, 'layer...","[{'size': '120x200', 'thickness': '5', 'price'...",Cao su thiên nhiên,Cao su,180.0
1,Nệm lò xo Amando Elite Original túi độc lập ti...,"https://cdn-cf-ops.vuanem.com/186x186,webp,q80...",https://vuanem.com/nem-lo-xo-amando-elite-orig...,Amando,26658.0,5.0,273.0,Thương hiệu Amando – Nghệ thuật giấc ngủ châu ...,"{'origin': 'Việt Nam', 'warranty': 120, 'layer...","[{'size': '120x200', 'thickness': '23', 'price...",Nệm Hybrid (Foam + Lò xo túi độc lập),Hybrid,120.0
2,Nệm foam Comfy Cloud 2.0 siêu đàn hồi dày 15cm,"https://cdn-cf-ops.vuanem.com/186x186,webp,q80...",https://vuanem.com/nem-foam-comfy-cloud-20.htm...,Comfy,25068.0,4.9,278.0,Comfy – Khoa học giấc ngủ cho cuộc sống cân bằ...,"{'origin': 'Việt Nam', 'warranty': 120, 'layer...","[{'size': '120x200', 'thickness': '15', 'price...",Nệm Hybrid (Cao su thiên nhiên + Foam),Hybrid,120.0
3,Nệm foam công nghệ Đức Goodnight Active Hybrid...,"https://cdn-cf-ops.vuanem.com/186x186,webp,q80...",https://vuanem.com/nem-foam-goodnight-active-h...,Goodnight,23094.0,5.0,139.0,Goodnight Active Hybrid - Chiếc nệm ĐẦU TIÊN ở...,"{'origin': 'Việt Nam', 'warranty': 120, 'layer...","[{'size': '120x200', 'thickness': '20', 'price...",Nệm Hybrid (Cao su thiên nhiên + Foam),Hybrid,120.0
4,Nệm lò xo Goodnight SleepWave Hybrid (Sleep Wa...,"https://cdn-cf-ops.vuanem.com/186x186,webp,q80...",https://vuanem.com/nem-lo-xo-goodnight-sleepwa...,Goodnight,22162.0,4.7,176.0,Goodnight SleepWave Hybrid – Công nghệ Hybrid ...,"{'origin': 'Việt Nam', 'warranty': 120, 'layer...","[{'size': '120x200', 'thickness': '20', 'price...",Nệm Hybrid (Foam + Lò xo túi độc lập),Hybrid,120.0


In [28]:
df_flat = flatten_dataset(mattress_dataset)
check_null_rate(df_flat)

Shape: (4532, 18)

=== NULL RATE (%) ===
technology             93.9
layer_composition      62.3
product_sold_number    41.5
reviews                41.5
rating                 41.5
firmness                3.0
warranty                1.8
brand                   0.1
origin                  0.1
product_name            0.0
link                    0.0
image_url               0.0
category                0.0
description             0.0
material_type           0.0
size                    0.0
thickness               0.0
price                   0.0
dtype: float64


## 📊 Kết quả kiểm tra Null Rate sau Flatten

### Thông tin cơ bản
- **Tổng số dòng:** 4,467
- **Tổng số cột:** 18
- **Nguồn dữ liệu:** 4 trang web (vuanem, khonemtonghop, thegioinem, thegioigiuongnem)

---

### Phân loại null rate từng cột

| Cột | Null Rate | Đánh giá | Hướng xử lý |
|-----|-----------|----------|-------------|
| `product_name` | 0.0% | ✅ Hoàn hảo | Giữ nguyên |
| `brand` | 0.1% | ✅ Tốt | Giữ nguyên |
| `material_type` | 0.0% | ✅ Hoàn hảo | Giữ nguyên |
| `category` | 0.0% | ✅ Hoàn hảo | Giữ nguyên |
| `description` | 0.0% | ✅ Hoàn hảo | Giữ nguyên |
| `link` | 0.0% | ✅ Hoàn hảo | Giữ nguyên |
| `image_url` | 0.0% | ✅ Hoàn hảo | Giữ nguyên |
| `origin` | 0.1% | ✅ Tốt | Giữ nguyên |
| `size` | 0.0% | ✅ Tốt | Giữ nguyên |
| `price` | 0.0% | ✅ Tốt | Giữ nguyên |
| `thickness` | 0.0% | ✅ Tốt | Giữ nguyên |
| `warranty` | 1.8% | ✅ Chấp nhận được | Giữ nguyên |
| `firmness` | 3.0% | ✅ Chấp nhận được | Impute theo kiến thức chuyên ngành |
| `rating` | 41.5% | ⚠️ Cao | Dùng tính popularity_score, không impute |
| `reviews` | 41.5% | ⚠️ Cao | Dùng tính popularity_score, không impute |
| `product_sold_number` | 41.5% | ⚠️ Cao | Dùng tính popularity_score, không impute |
| `layer_composition` | 62.3% | ❌ Quá nhiều | Không dùng làm feature model |
| `technology` | 93.9% | ❌ Quá nhiều | Drop khỏi dataset |

---

### Nhận xét

**Nguyên nhân null 41.5% ở rating/reviews/product_sold_number:**
- Toàn bộ đến từ **thegioigiuongnem.com** — trang này không hiển thị thông tin xã hội
- Các trang còn lại (vuanem, khonemtonghop, thegioinem) đều có đủ 3 cột này

**Nguyên nhân null cao ở layer_composition (62.3%) và technology (93.9%):**
- Đây là thông tin kỹ thuật chuyên sâu, không phải trang nào cũng công bố
- Không đủ độ tin cậy để đưa vào model

**Các cột quan trọng cho model đều sạch:**
- `product_name`, `brand`, `material_type`, `category`, `description` → 0% null
- Đây là các feature chính của Content-Based Filtering

---

In [29]:
df_flat.shape
print(df_flat.columns.tolist())

['product_name', 'brand', 'material_type', 'category', 'description', 'rating', 'reviews', 'product_sold_number', 'image_url', 'link', 'origin', 'warranty', 'firmness', 'layer_composition', 'technology', 'size', 'thickness', 'price']


In [30]:
df_flat = df_flat.drop(columns=['layer_composition', 'technology'])

print(df_flat.shape)       
print(df_flat.columns.tolist())

(4532, 16)
['product_name', 'brand', 'material_type', 'category', 'description', 'rating', 'reviews', 'product_sold_number', 'image_url', 'link', 'origin', 'warranty', 'firmness', 'size', 'thickness', 'price']


In [31]:
print(df_flat.dtypes)

product_name            object
brand                   object
material_type           object
category                object
description             object
rating                 float64
reviews                float64
product_sold_number    float64
image_url               object
link                    object
origin                  object
warranty                object
firmness                object
size                    object
thickness               object
price                    int64
dtype: object


In [32]:
df_flat['warranty'] = df_flat['warranty'].apply(parse_warranty)

# Kiểm tra
print(df_flat['warranty'].value_counts().sort_index())
print(df_flat['warranty'].describe())

warranty
12.0       15
36.0       41
60.0      632
72.0       50
84.0      241
96.0      302
108.0      21
120.0    2035
144.0     732
180.0     301
216.0      25
240.0      24
Name: count, dtype: int64
count    4419.000000
mean      115.326544
std        34.729387
min        12.000000
25%        96.000000
50%       120.000000
75%       120.000000
max       240.000000
Name: warranty, dtype: float64


In [33]:
df_flat[['width', 'length']] = df_flat['size'].apply(
    lambda x: pd.Series(parse_size(x))
)
df_flat['thickness'] = df_flat['thickness'].apply(parse_thickness)

# Xoá cột size gốc
df_flat = df_flat.drop(columns=['size'])

# Kiểm tra
print(df_flat[['width', 'length', 'thickness', 'warranty']].dtypes)
print(df_flat[['width', 'length', 'thickness', 'warranty']].describe())

width        float64
length       float64
thickness    float64
warranty     float64
dtype: object
             width       length    thickness     warranty
count  4525.000000  4525.000000  4521.000000  4419.000000
mean    153.807293   199.153812    15.129308   115.326544
std      37.535123    10.814047     8.027049    34.729387
min      16.000000    35.000000     2.000000    12.000000
25%     120.000000   200.000000    10.000000    96.000000
50%     160.000000   200.000000    14.000000   120.000000
75%     180.000000   200.000000    20.000000   120.000000
max     240.000000   240.000000    39.000000   240.000000


In [34]:
# Xem các giá trị unique
print(df_flat['firmness'].value_counts())

firmness
Trung bình                                                                                1440
Cứng                                                                                       957
Cứng vừa                                                                                   500
Mềm vừa                                                                                    439
Trung bình - Hơi cứng                                                                      324
Mềm - Trung bình                                                                           193
Mềm                                                                                        161
Cứng trung bình (Vững)                                                                     122
Cứng (Vững chắc)                                                                           100
Trung bình (Cân bằng)                                                                       64
Trung Bình                               

In [35]:
df_flat['firmness'] = df_flat['firmness'].apply(
    normalize_firmness
)
print(df_flat['firmness'].value_counts())

firmness
Trung bình               1549
Cứng                      986
Cứng vừa                  726
Mềm - Trung bình          644
Trung bình - Hơi cứng     324
Mềm                       161
Name: count, dtype: int64


In [36]:
print(df_flat['origin'].value_counts())

origin
Việt Nam                               3544
Anh Quốc                                216
Hàn Quốc                                162
Nệm Kim Cương - Việt Nam                148
Hoa Kỳ                                   91
Nệm Ưu Việt - Việt Nam                   72
Công ty Nệm ACB                          41
Mỹ                                       34
Nhật Bản                                 25
TATANA - Việt Nam                        23
Nệm Kim Cương                            22
Cty nệm Kim Cương - Việt Nam             20
Anh                                      19
Anh quốc                                 18
Công ty nệm King Koil                    17
VIệt Nam                                 16
Thái Lan                                 15
Nệm Kim Cương - Việt Nam                 15
Viết Nam                                  7
Công ty Nệm King Koil                     5
Liên Á                                    4
 Artemis - Everon                         3
Artemis - Everon         

In [37]:
df_flat['origin'] = df_flat['origin'].apply(standardize_origin)
print(df_flat['origin'].value_counts())

origin
Việt Nam    3915
Anh Quốc     253
Hàn Quốc     171
Mỹ           147
Nhật Bản      25
Thái Lan      15
Name: count, dtype: int64


## Chuẩn hoá các cột số và Firmness

### 1. Parse Size → Width & Length

**Vấn đề:** Cột `size` lưu dạng string `"160x200"`, `"160 x 200"` 
→ Không thể dùng trực tiếp cho model

**Xử lý:** Tách thành 2 cột số riêng biệt
- `"160x200"` → `width=160.0`, `length=200.0`
- Không parse được → `None`

---

### 2. Parse Thickness

**Vấn đề:** Cột `thickness` lưu dạng string `"10cm"`, `"10"`, `"10.5cm"`

**Xử lý:** Trích xuất số, bỏ đơn vị
- `"10cm"` → `10.0`
- `"10.5"` → `10.5`
- Không parse được → `None`

---

### 3. Chuẩn hoá Firmness

**Vấn đề:** Cột `firmness` có 14 giá trị khác nhau do:
- Viết hoa/thường không nhất quán: `"Trung bình"` vs `"Trung Bình"`
- Diễn đạt khác nhau cùng nghĩa: `"Cứng (Vững chắc)"` = `"Cứng vừa"`
- Description bị crawl nhầm vào firmness (chuỗi > 50 ký tự)
- Giá trị mơ hồ: `"Cao"` → map về `"Cứng"`

**Mapping chuẩn hoá về 7 nhóm:**

| Giá trị gốc | Sau chuẩn hoá |
|-------------|---------------|
| Trung bình, Trung Bình, Trung bình (Vững), Trung bình (Cân bằng), Độ cứng trung bình | `Trung bình` |
| Cứng vừa, Cứng trung bình, Cứng (Vững chắc) | `Cứng vừa` |
| Trung bình - Hơi cứng | `Trung bình - Hơi cứng` |
| Mềm - Trung bình, Mềm vừa, Mềm trung bình (Êm ái) | `Mềm - Trung bình` |
| Cứng, Cao | `Cứng` |
| Mềm | `Mềm` |
| Chuỗi > 50 ký tự (description nhầm) | `None` |

**Kết quả:** Từ 14 giá trị → 7 nhóm chuẩn

### 4. Chuẩn hoá origin 

**Vấn đề:** Cột `origin` có nhiều giá trị sai nguồn gốc:
- Mỹ, Hoa Kỳ
- Everon là hãng đến từ Hàn Quốc
- Còn lại, thay vì để là Việt Nam thì lại hiện tên của công ty sản xuất nó

In [38]:
df_flat['description'] = df_flat['description'].apply(clean_text)

In [39]:
df_flat.head()

,product_name,brand,material_type,category,description,rating,reviews,product_sold_number,image_url,link,origin,warranty,firmness,thickness,price,width,length
0,Nệm cao su thiên nhiên Gummi Classic thế hệ mớ...,Gummi,Cao su thiên nhiên,Cao su,"Thương hiệu Gummi - Tinh khiết từ cao su, đúng...",4.9,359.0,44410.0,"https://cdn-cf-ops.vuanem.com/186x186,webp,q80...",https://vuanem.com/nem-cao-su-gummi-classic.ht...,Việt Nam,180.0,Cứng vừa,5.0,6712000,120.0,200.0
1,Nệm cao su thiên nhiên Gummi Classic thế hệ mớ...,Gummi,Cao su thiên nhiên,Cao su,"Thương hiệu Gummi - Tinh khiết từ cao su, đúng...",4.9,359.0,44410.0,"https://cdn-cf-ops.vuanem.com/186x186,webp,q80...",https://vuanem.com/nem-cao-su-gummi-classic.ht...,Việt Nam,180.0,Cứng vừa,10.0,7808000,120.0,200.0
2,Nệm cao su thiên nhiên Gummi Classic thế hệ mớ...,Gummi,Cao su thiên nhiên,Cao su,"Thương hiệu Gummi - Tinh khiết từ cao su, đúng...",4.9,359.0,44410.0,"https://cdn-cf-ops.vuanem.com/186x186,webp,q80...",https://vuanem.com/nem-cao-su-gummi-classic.ht...,Việt Nam,180.0,Cứng vừa,15.0,12608000,120.0,200.0
3,Nệm cao su thiên nhiên Gummi Classic thế hệ mớ...,Gummi,Cao su thiên nhiên,Cao su,"Thương hiệu Gummi - Tinh khiết từ cao su, đúng...",4.9,359.0,44410.0,"https://cdn-cf-ops.vuanem.com/186x186,webp,q80...",https://vuanem.com/nem-cao-su-gummi-classic.ht...,Việt Nam,180.0,Cứng vừa,5.0,7352000,140.0,200.0
4,Nệm cao su thiên nhiên Gummi Classic thế hệ mớ...,Gummi,Cao su thiên nhiên,Cao su,"Thương hiệu Gummi - Tinh khiết từ cao su, đúng...",4.9,359.0,44410.0,"https://cdn-cf-ops.vuanem.com/186x186,webp,q80...",https://vuanem.com/nem-cao-su-gummi-classic.ht...,Việt Nam,180.0,Cứng vừa,10.0,9408000,140.0,200.0


In [40]:
df_flat.dtypes

product_name            object
brand                   object
material_type           object
category                object
description             object
rating                 float64
reviews                float64
product_sold_number    float64
image_url               object
link                    object
origin                  object
warranty               float64
firmness                object
thickness              float64
price                    int64
width                  float64
length                 float64
dtype: object

In [41]:
# Lưu processed dataset
df_flat.to_csv(
    "../data/processed/mattresses_clean.csv",
    index=False,
    encoding='utf-8-sig'  
)

file_path = "../data/processed/mattresses_clean.csv"
file_size = os.path.getsize(file_path) / (1024 * 1024)

print(f"Đã lưu: {file_path}")
print(f"Kích thước: {file_size:.2f} MB")
print(f"Shape: {df_flat.shape}")
print(f"Columns: {df_flat.columns.tolist()}")

Đã lưu: ../data/processed/mattresses_clean.csv
Kích thước: 18.33 MB
Shape: (4532, 17)
Columns: ['product_name', 'brand', 'material_type', 'category', 'description', 'rating', 'reviews', 'product_sold_number', 'image_url', 'link', 'origin', 'warranty', 'firmness', 'thickness', 'price', 'width', 'length']
